### ISSUE TITLE
unknown_limit in 2.0.65

### ISSUE THREAD

**Issue by kimheino** [munin-monitoring/munin#1361]
I have warning emails enabled and tested different values for unknown_limit option.

Steps to reproduce: Modify any of the existing plugin to have warning limit "1:6" and hardcode it to return value "1" or "U".

With "unknown_limit 0" I get warning email for 1st U value (time +5min).
With "unknown_limit 1" I get warning email for 1st U value (time +5min).
With "unknown_limit 2" I get warning email for 3rd U value (time +15min).
With "unknown_limit 3" I get warning email for 4th U value (time +20min).

How can I get warning for 2nd U value? I mean "value 1", 5min later "U", 5min later "U" => warning email. I assume unknown_limit 1 or 2 should do that.

**Comment by kimheino** [comment]
I think it would be easy to fix "unknown_limit 1" to be "2nd U sends warning email". Would this be acceptable change?

**Comment by sumpfralle** [comment]
Would this change fix the issue for you?

```perl
diff --git a/master/lib/Munin/Master/LimitsOld.pm b/master/lib/Munin/Master/LimitsOld.pm
index 643873ef7..3ce550090 100644
--- a/master/lib/Munin/Master/LimitsOld.pm
+++ b/master/lib/Munin/Master/LimitsOld.pm
@@ -441,7 +441,7 @@ sub process_service {
             # First we'll need to check whether the user wants to ignore
             # a few UNKNOWN values before actually changing the state to
             # UNKNOWN.
-            if (($oldstate ne "unknown") and ($unknown_limit > 1)) {
+            if (($oldstate ne "unknown") and ($unknown_limit >= 1)) {
                  if (!defined($onfield->{"num_unknowns"}) or ($onfield->{"num_unknowns"} < $unknown_limit)) {
                      $newstate = $oldstate;
                      $extinfo = $onfield->{$newstate};
```

(it is a bit of a wild guess - this status handling code flow is a bit hard to read)

**Comment by kimheino** [comment]
That fixes "unknown_limit 1", but it's only part of the fix. See lines 595-598. Those lines should be fixed to allow zero value, negatives being not valid. But then again, regexp in line 592 doesn't allow negative numbers, so simply delete lines 594-599:

```
    if ($unknown_limit =~ /^\s*(\d+)\s*$/) {                                    
        $unknown_limit = $1 if defined $1;                                      
        DEBUG "[DEBUG] processing unknown_limit: $name -> $unknown_limit";      
    }   
```

My perl-fu is not powerful, why is "if defined $1" needed here after strict regexp?

**Comment by sumpfralle** [comment]
> hat fixes "unknown_limit 1", but it's only part of the fix. See lines 595-598. Those lines should be fixed to allow zero value, negatives being not valid. But then again, regexp in line 592 doesn't allow negative numbers, so simply delete lines 594-599:

Yes, now that makes sense. The code flow is really a bit twisted :(
Thank you for your input - I just pushed the fix to `stable-2.0`.

> My perl-fu is not powerful, why is "if defined $1" needed here after strict regexp?

`if defined` was introduced in baef89ee5b98ca35021116374865f4dcf1c40e4f. Just as you, I cannot imagine, why it is necessary. I removed it now.
